clusterloader2/pkg/measurement/common/service_creation_latency.go

```diff
@@ -0,0 +1,411 @@
+	perfData.DataItems = append(perfData.DataItems, latency.CreateToAvailableLatency.ToPerfData("create_to_available"))
+	return perfData
+}
+
+func (s *serviceCreationLatencyMeasurement) handleObject(oldObj, newObj interface{}) {
+	var oldService *corev1.Service
+	var newService *corev1.Service
+	var ok bool
+	oldService, ok = oldObj.(*corev1.Service)
+	if oldObj != nil && !ok {
+		klog.Errorf("%s: uncastable old object: %v", s, oldObj)
+		return
+	}
+	newService, ok = newObj.(*corev1.Service)
+	if newObj != nil && !ok {
+		klog.Errorf("%s: uncastable new object: %v", s, newObj)
+		return
+	}
+	isEqual := oldService != nil &&
+		newService != nil &&
+		equality.Semantic.DeepEqual(oldService.Spec, newService.Spec) &&
+		equality.Semantic.DeepEqual(oldService.Status, newService.Status)
+	if isEqual {
+		// Skip updates without changes in the spec.
+		return
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if !s.isRunning {
+		return
+	}
+	if newObj == nil {
+		if err := s.deleteObjectLocked(oldService); err != nil {
+			klog.Errorf("%s: delete checker error: %v", s, err)
+		}
+		return
+	}
+	if err := s.updateObjectLocked(newService); err != nil {
+		klog.Errorf("%s: create checker error: %v", s, err)
+	}
+}
+
+func (s *serviceCreationLatencyMeasurement) deleteObjectLocked(svc *corev1.Service) error {
+	key, err := runtimeobjects.CreateMetaNamespaceKey(svc)
+	if err != nil {
+		return fmt.Errorf("meta key created error: %v", err)
+	}
+	if pc, exists := s.pingCheckers[key]; exists {
+		pc.stop()
+		delete(s.pingCheckers, key)
+	}
+	delete(s.creationTimes, key)
+	return nil
+}
+
+func (s *serviceCreationLatencyMeasurement) updateObjectLocked(svc *corev1.Service) error {
+	key, err := runtimeobjects.CreateMetaNamespaceKey(svc)
+	if err != nil {
+		return fmt.Errorf("meta key created error: %v", err)
+	}
+	if _, exists := s.creationTimes[key]; !exists {
+		s.creationTimes[key] = &serviceCreationTimes{
+			created: time.Now(),
+		}
+	}
+	if len(svc.Status.LoadBalancer.Ingress) < 1 {
+		return nil
+	}
+	ct := s.creationTimes[key]
+	if ct.ipAssigned.IsZero() {
+		klog.Infof("%s: %s ipAssigned", s, key)
+		ct.ipAssigned = time.Now()
+	}
+	if _, exists := s.pingCheckers[key]; !exists {
+		s.pingCheckers[key] = &pingChecker{
+			caller:        s.String(),
+			key:           key,
+			svc:           svc,
+			creationTimes: ct,
+			stopCh:        make(chan struct{}),
+		}
+		go s.pingCheckers[key].run()
+	}
+	return nil
+}
+
+type pingChecker struct {
+	caller        string
+	key           string
+	svc           *corev1.Service
+	creationTimes *serviceCreationTimes
+	stopCh        chan struct{}
+}
+
+func (p *pingChecker) run() {
+	success := 0
+	for {
+		select {
+		case <-p.stopCh:
+			return
+		default:
+			if !p.creationTimes.reachable.IsZero() {
+				return
+			}
+			if len(p.svc.Status.LoadBalancer.Ingress) < 1 || len(p.svc.Spec.Ports) < 1 {
```

Comment:
This isn't needed - you're already checking that in updateObjectLocked.

The contract of run() function should be that it takes only services that are pingable.

Reply:
Logging moved from pingChecker to service update handler.